### PR TITLE
End game condition check

### LINKED ISSUE #682: A game is over when the time limit is reached
### Depends

- [ ] #681

### Context

The game needs to be able to check itself for when it is time to finish. The game should terminate when the game hits the time limit.

### Test
- [x] Test is ready and on develop 

```cpp
//#define FIX_ISSUE_682
    #ifdef FIX_ISSUE_682
    // (682) A game is over when the time limit is exceeded
    {
      const int time_limit = 10;
      const game_options g_options{3,
                            false,
                            get_random_kam(),
                            get_random_kam(),
                            get_random_kam(),
                            environment_type(),
                            time_limit
                           };
      game g{g_options};
      for (int i = 0; i < time_limit; i++)
        {
          assert(!g.is_over());
          g.tick();
        }
      assert(g.is_over());
    }
    #endif // FIX_ISSUE_682
```

### PR DESCRIPTION
**Pull request by EvoLandEco**
fixes #682

### CODE REVIEW

File: src/game_options.cpp [hunk 1]
```diff
@@ -188,7 +190,7 @@ void test_game_options()
           environment_type(),
           max_time
           );
-    assert(game_options.get_max_time() == max_time);
+    assert(random_game_options.get_max_time() == max_time);
```
Comment (TheoPannetier) [inline src/game_options.cpp:193]: Oops, thanks for catching this mistake in the test!

File: src/game_options.cpp [hunk 2]
```diff
@@ -170,11 +172,11 @@ void test_game_options()
     game_options go;
     assert(go.get_environment_type() == environment_type::empty);
   }
-//#define FIX_ISSUE_681
+#define FIX_ISSUE_681
 #ifdef FIX_ISSUE_681
   // (681) The game_options set a time limit
   {
-    const int max_time = 500; // change if too long/short
+    const int max_time = 6000; // change if too long/short
```
Comment (TheoPannetier) [inline src/game_options.cpp:179]: That looks appropriate :)

File: src/game.cpp [hunk 3]
```diff
@@ -1172,8 +1182,29 @@ void test_game() //!OCLINT tests may be many
...
+    game g{g_options};
+    for (int i = 0; i < time_limit; i++)
+      {
+        assert(!g.is_over());
+        g.increment_n_ticks();
+        g.check_over();
+      }
+    assert(g.is_over());
+  }
+
```
Comment (TheoPannetier) [inline src/game.cpp:1207]: I feel this test block is redundant with the next one. We don't need to explicitly test that `check_over()` exists: its existence (or something similar to it) is anyways required to make the test block below pass.

```suggestion
```

File: src/game.cpp [hunk 4]
```diff
@@ -217,6 +220,13 @@ void game::reset_player_action()
     }
 }
 
+/// END GAME CONDITION ///
+void game::check_over()
+{
+  if (m_n_ticks >= m_options.get_max_time()) m_is_over = true;
```
Comment (TheoPannetier) [inline src/game.cpp:226]: This is good, but you could make it even more concise if you like ;)

**Review by TheoPannetier — APPROVED** [review]
Excellent solution! I've made a suggestion to remove the extra test you have added (I feel it's redundant with the existing one), but feel free to disagree and merge as is.

**Comment by EvoLandEco** [comment]
> Excellent solution! I've made a suggestion to remove the extra test you have added (I feel it's redundant with the existing one), but feel free to disagree and merge as is.

I was concerning that we might change the way we implement things later, an example is that we either implement a function outsides a class, or as a public method, I think it would be even safer to have the test to ensure that we do not accidentally change the exposed API

File: src/game_options.cpp (hunk 2 above)
Comment (EvoLandEco) [reply src/game_options.cpp:179]: Might still be too short for one round, but that's not a thing we should think about for now

**Pull request merged**